packages/grid/_modules_/grid/hooks/features/useComponents.tsx

```diff
@@ -69,5 +68,11 @@ export const useComponents = (
     };
   }, [baseComponentProps, componentsProp]);
 
+  React.useEffect(() => {
+    if (apiRef && apiRef.current) {
+      apiRef.current.components = components;
+    }
+  }, [apiRef, components]);
```

Comment:
This looks wrong. We update the apiRef in an effect, and yet, in the render, we read directly from it. No way it can work.

Reply:
Why not use the context to propagate the components? The fewer we rely on global side effects (apiRef), the better.

Reply:
Yeah bingo, see "bug 1" in https://codesandbox.io/s/material-demo-forked-m7ilt?file=/demo.js. It only update after we resize.

Reply:
Fixed https://codesandbox.io/s/create-react-app-forked-elq43?file=/src/App.js

Reply:
By putting it on ApiRef it will be accessible anywhere. This allows us to render custom components from anywhere in the react tree. 
Also by putting everything on ApiRef, users would have access to any prop using one single object as one single source of truth thus making the learning curve pretty easy.
Before I added multiple contexts, for rendering props and options, but I refactored as it's much easier to attach it to apiRef and pass it around.

Reply:
This fixes the reproduction.

However, it also violates one of the rendering principles of React. React can render something and never commit it (update the DOM). This mean that renders should be side effect free. This is the opposite of what we do. We update a ref during a render, a state that can corrupt the next render. This is the mistake we/I did in `@material-ui/styles` and why it's not working in StrictMode, why we can't enable StrictMode in the docs for v4.

I think that this will likely require a re-architecture of the apiRef and state, it might come with #849. But for now, it's working so probably good enough, I also don't expect it to require a large refactoring in the future. It should be relatively straightforward.

> Also by putting everything on ApiRef, users would have access to any prop using one single object as one single source of truth thus making the learning curve pretty easy.

I don't understand the value for the developers and I'm worried about the cost for them.
From what I understand, **by default!**, we are making most of the internals of the data grid public. I have always considered the opposite to be a better strategy: make as much as we can private, and force developers to come up with problems so we can evaluate how best to solve them, expose public APIs when needed.
Will we be able to deliver large improvements in the data grid once we leave the alpha phase without requiring breaking changes, crippling our ability to deliver value?